planner/core/optimizer.go

```diff
@@ -687,31 +695,183 @@ const (
 	unknown shuffleTarget = iota
 	window
 	joinBuild
+	hashAgg
 )
 
 type fineGrainedShuffleHelper struct {
 	shuffleTarget shuffleTarget
 	plans         []*basePhysicalPlan
+	joinKeysCount int
+}
+
+type tiflashClusterInfoStatus uint8
+
+const (
+	unInitialized tiflashClusterInfoStatus = iota
+	initialized
+	failed
+)
+
+type tiflashClusterInfo struct {
+	itemStatus tiflashClusterInfoStatus
+	itemValue  uint64
 }
 
 func (h *fineGrainedShuffleHelper) clear() {
 	h.shuffleTarget = unknown
 	h.plans = h.plans[:0]
+	h.joinKeysCount = 0
 }
 
 func (h *fineGrainedShuffleHelper) updateTarget(t shuffleTarget, p *basePhysicalPlan) {
 	h.shuffleTarget = t
 	h.plans = append(h.plans, p)
 }
 
-func setupFineGrainedShuffleInternal(plan PhysicalPlan, helper *fineGrainedShuffleHelper, streamCount uint64) {
+// calculateTiFlashStreamCountUsingMinLogicalCores uses minimal logical cpu cores among tiflash servers, and divide by 2
+// return false, 0 if any err happens
+func calculateTiFlashStreamCountUsingMinLogicalCores(ctx context.Context, sctx sessionctx.Context, serversInfo []infoschema.ServerInfo) (bool, uint64) {
+	failpoint.Inject("mockTiFlashStreamCountUsingMinLogicalCores", func(val failpoint.Value) {
+		intVal, err := strconv.Atoi(val.(string))
+		if err == nil {
+			failpoint.Return(true, uint64(intVal))
+		} else {
+			failpoint.Return(false, 0)
+		}
+	})
+	rows, err := infoschema.FetchClusterServerInfoWithoutPrivilegeCheck(ctx, sctx, serversInfo, diagnosticspb.ServerInfoType_HardwareInfo, false)
+	if err != nil {
+		return false, 0
+	}
+	var initialMaxCores uint64 = 10000
+	var minLogicalCores uint64 = initialMaxCores // set to a large enough value here
+	for _, row := range rows {
+		if row[4].GetString() == "cpu-logical-cores" {
+			logicalCpus, err := strconv.Atoi(row[5].GetString())
+			if err == nil && logicalCpus > 0 && uint64(logicalCpus) < minLogicalCores {
+				minLogicalCores = uint64(logicalCpus)
+			}
+		}
+	}
+	// No need to check len(serersInfo) == serverCount here, since missing some servers' info won't affect the correctness
+	if minLogicalCores > 1 && minLogicalCores != initialMaxCores {
+		return true, minLogicalCores / 2
+	}
+
+	return false, 0
+}
+
+func checkFineGrainedShuffleForJoinAgg(ctx context.Context, sctx sessionctx.Context, streamCountInfo *tiflashClusterInfo, tiflashServerCountInfo *tiflashClusterInfo, exchangeColCount int, splitLimit uint64) (applyFlag bool, streamCount uint64) {
+	switch (*streamCountInfo).itemStatus {
+	case unInitialized:
+		streamCount = 4 // assume 8c node in cluster as minimal, stream count is 8 / 2 = 4
+	case initialized:
+		streamCount = (*streamCountInfo).itemValue
+	case failed:
+		return false, 0 // probably won't reach this path
+	}
+
+	var tiflashServerCount uint64 = 0
+	switch (*tiflashServerCountInfo).itemStatus {
+	case unInitialized:
+		serversInfo, err := infoschema.GetTiFlashServerInfo(sctx)
```

Comment:
We may also consider disaggregated tiflash mode, the label of tiflash_compute nodes is `<engine, tiflash_compute>`

Reply:
Good point, disable this for disaggregated tiflash mode by return unsupported error.

Reply:
And we can also init `streamCountInfo` here?

Reply:
Done